### ISSUE TITLE
git_commit_summary can return newlines which leads to invalid reflog messages

### ISSUE THREAD

**Issue by srajko** [libgit2/libgit2#3458]
Newlines after spaces present a problem to `git_commit_summary`.  For example, `git_commit_summary("Newline after space \nis removed")` returns:
```
"Newline after space  \nis removed"
```

Since `git_reference__update_for_commit` uses `git_commit_summary` to construct the reflog message, the returned newline in the reflog message breaks `git_reflog_read` (it will return an error because it finds the newline, expects a new entry, and then can't parse the oid).

I'd be happy to submit a PR if fixing `git_commit_summary` is indeed the right thing to do here.

**Comment by ethomson** [comment]
Yes!  It certainly looks like `git_commit_summary` is deficient here - a pull request would be great!

**Comment by carlosmn** [comment]
It doesn't seem like `git_commit_summary()` is deficient here, but rather that we're not handling it correctly. The summary is a paragraph and can thus contain many newlines. What does git do in this case? Does it reflow the text to make it a single line or does it cut at the first newline?

**Comment by srajko** [comment]
`git_commit_summary` does process the first paragraph, however it tries to remove newlines and return a single line result.  Here are the original test cases:

```
	assert_commit_summary("One-liner with no trailing newline", "One-liner with no trailing newline");
	assert_commit_summary("One-liner with trailing newline", "One-liner with trailing newline\n");
	assert_commit_summary("Trimmed leading&trailing newlines", "\n\nTrimmed leading&trailing newlines\n\n");
	assert_commit_summary("First paragraph only", "\nFirst paragraph only\n\n(There are more!)");
	assert_commit_summary("First paragraph with  unwrapped trailing\tlines", "\nFirst paragraph\nwith  unwrapped\ntrailing\tlines\n\n(Yes, unwrapped!)");
	assert_commit_summary("\tLeading \ttabs", "\tLeading\n\ttabs\n\nis preserved");
	assert_commit_summary(" Leading  Spaces", " Leading\n Spaces\n\nare preserved");
	assert_commit_summary("Trailing tabs\tare removed", "Trailing tabs\tare removed\t\t");
	assert_commit_summary("Trailing spaces  are removed", "Trailing spaces  are removed  ");
	assert_commit_summary("Trailing tabs", "Trailing tabs\t\n\nare removed");
	assert_commit_summary("Trailing spaces", "Trailing spaces \n\nare removed");
	assert_commit_summary("", "");
	assert_commit_summary("", " ");
	assert_commit_summary("", "\n");
	assert_commit_summary("", "\n \n");
```
and a couple I added (note both of these pass):
```
	assert_commit_summary("Newline is removed", "Newline\nis removed");
	assert_commit_summary("Newline  \nis removed", "Newline \nis removed");;
```
The fact that `"Newline\nis removed"` gets the newline removed but `"Newline \nis removed"` doesn't seems like a red herring to me.

As far as what git does In this case - it cuts at the first newline to make the reflog message.

It's also worth noting that `git reflog` manages to output the reflog after a newline-inside-reflog-message has been added to it - it will just throw out the part of the message following the newline (but if you have it rewrite the reflog, it will rewrite it with the extraneous message line thrown out).

**Comment by srajko** [comment]
There is also a related issue in `reflog_parse`, leading to a potential segfault when `git_reflog_read` runs into a reflog that has been corrupted by such an extraneous newline.  Here is the relevant code:

```
		entry = git__calloc(1, sizeof(git_reflog_entry));
		GITERR_CHECK_ALLOC(entry);

		entry->committer = git__malloc(sizeof(git_signature));
		GITERR_CHECK_ALLOC(entry->committer);

		if (git_oid_fromstrn(&entry->oid_old, buf, GIT_OID_HEXSZ) < 0)
			goto fail;
...
fail:
	if (entry)
		git_reflog_entry__free(entry);
```

The extra newline causes this code to try to process the 2nd line of the reflog message thinking that it is a full reflog entry, expecting it to start with an oid.  `git_oid_fromstrn` returns an error, and in `fail:` we do `git_reflog_entry__free`.  This leads to `git_signature_free(entry->committer);`, which does `git__free(entry->committer->name);` and `git__free(entry->committer->email);`.

Because `entry->committer` was allocated via `git__malloc`, `entry->committer->name` and `entry->committer->email` are undefined and freeing them can segfault.  Using `git__calloc` might be safer - that or moving the allocation further down after the oids are parsed.

**Comment by carlosmn** [comment]
Fixed via #3486 and #3487
